File: Python/binary_search_01.py
# 2022.01.18 이진탐색
# 부품찾기

# 이진탐색 메소드
def sol(array, target, start, end):
    while (start <= end):
        mid = (start + end) // 2
        # 일치하는 경우 반환
        if array[mid] == target:
            return mid
        # mid 의 좌측 탐색
        elif array[mid] > target:
            end = mid - 1
        # mid 의 우측 탐색
        else:
            start = mid + 1
    return None     # 없으면 None 반환

# input n, m, list
n = int(input())
array = list(map(int, input().split()))
array.sort()

m = int(input())
m_list = list(map(int, input().split()))

# 요청 품목이 있을 경우 yes 출력
# 요청 품목이 없을 경우 no 출력
for i in m_list:
    ans = sol(array, i, 0, n-1)
    if ans != None:
        print('yes', end=' ')
    else:
        print('no', end=' ')

File: Java/CodingTest/src/05_BinarySearch/BinarySearch_01.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

// 부품 찾기
public class BinarySearch_01 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        // N 과 N 개의 부품 입력
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int[] items = new int[n];
        for (int i = 0; i < n; i++) {
            items[i] = Integer.parseInt(st.nextToken());
        }
        // items 정렬
        Arrays.sort(items);

        // M 입력
        int m = Integer.parseInt(br.readLine());
        // M 개의 부품 입력과 동시에 이진 탐색 수행
        st = new StringTokenizer(br.readLine(), " ");
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < m; i++) {
            int target = Integer.parseInt(st.nextToken());
            // 찾으면 "yes" 찾지 못하면 "no" 를 StringBuilder 에 추가
            if (binarySearch(items, target, 0, n-1) == -1) {
                sb.append("no").append(" ");
            }
            else {
                sb.append("yes").append(" ");
            }
        }
        System.out.println(sb);
    }
    
    // 이진 탐색 메소드
    public static int binarySearch(int[] arr, int target, int start, int end) {
        while (start <= end) {
            int mid = (int)((start + end) / 2);
            // 중간점의 값이 찾고자 하는 값과 같을 경우 중간점의 인덱스 반환
            if (arr[mid] == target) {
                return mid;
            }
            // 좌측 탐색
            else if (arr[mid] > target) {
                end = mid - 1;
            }
            // 우측 탐색
            else {
                start = mid + 1;
            }
        }
        // 못찾을 경우 -1 반환
        return -1;
    }
}
